04-onion-routing.md

```diff
@@ -150,52 +155,75 @@ The overall structure of the packet is as follows:
+Additionally, `hop_payload`s are incrementally obfuscated at each hop.
 
-The `realm` byte determines the format of the `per_hop` field; currently, only `realm`
-0 is defined, for which the `per_hop` format follows:
+Each `hop_payload` has the following structure:
 
-1. type: `per_hop` (for `realm` 0)
+1. type: `hop_payload`
+2. data:
+   * [`1`: `num_frames_and_realm`]
+   * [`raw_payload_len`: `raw_payload`]
+   * [`padding_len`: `padding`]
+   * [`32`: `HMAC`]
+
+Notice that since the `hop_payload` is instantiated once per hop, the subscript `_i` may be used in the remainder of this document to refer to the `hop_payload` and its fields destined for hop `i`.
+
+The `hop_payload` consists of at least one `frame` followed by up to 15 additional `frame`s.
+The number of additional frames allocated to the current hop is determined by the 4 most significant bits of `num_frames_and_realm`, while the 4 least significant bits determine the payload format.
+Therefore the number of frames allocated to the current hop is given by `num_frames = (num_frames_and_realm >> 4) + 1`.
+For simplification we will use `hop_payload_len` to refer to `num_frames * FRAME_SIZE`.
+
+In order to have sufficient space to serialized the `raw_payload` into the `hop_payload` while minimizing the number of used frames the number of frames used for a single `hop_payload` MUST be equal to
+
+> `num_frames = ceil((raw_payload_len + 1 + 32) / FRAME_SIZE)`
+
+The payload format determines how the `raw_payload` should be interpreted (see below for currently defined formats), and how much padding is added.
+In order to position the `HMAC` in the last 32 bytes of the `hop` the `raw_payload` MUST be followed by `padding_len = (num_frames * FRAME_SIZE - 1 - raw_payload_len - 32)` `0x00`-bytes.
+
+The `realm` is specified as `num_frames_and_realm & 0x0F`.
+It determines the format of the `raw_payload` field; the following `realm`s are currently defined.
+
+| `realm` | Payload Format                                             |
+|:--------|:-----------------------------------------------------------|
+| `0x00`  | [Version 1 `hop_data`](#version-1-hop_data-payload-format) |
+| `0x01`  | TLV payload format (to be specified)                       |
+
+Using the `hop_payload` field, the origin node is able to precisely specify the path and structure of the HTLCs forwarded at each hop.
+As the `hop_payload` is protected under the packet-wide HMAC, the information it contains is fully authenticated with each pair-wise relationship between the HTLC sender (origin node) and each hop in the path.
+
+Using this end-to-end authentication, each hop is able to cross-check the HTLC parameters with the `hop_payload`'s specified values and to ensure that the sending peer hasn't forwarded an ill-crafted HTLC.
+
+## Version 1 `hop_data` Payload Format
+
+The version 1 `hop_data` payload format has realm `0x00`, and MUST use a single `frame` to encode the payload.
```

Comment:
Realm should completely re-defined above. Atm, it is hidden within a greater paragraph the 4 bit distinction.